File: android/src/main/kotlin/com/exelbid/flutter/mappers/WaterfallEventMapper.kt
package com.exelbid.flutter.mappers

/**
 * Builds the `"waterfall"` payload maps for `onWaterfall` events.
 *
 * Unlike iOS (where the SDK emits a typed `EBWaterfallEvent` we encode), the
 * Android plugin runs the waterfall itself, so these are constructed directly.
 * The shape is byte-for-byte identical to iOS `WaterfallEventMapper` so the Dart
 * `WaterfallEvent.fromMap` sealed class decodes both platforms unchanged.
 */
object WaterfallEventMapper {

    fun fetching(): Map<String, Any?> = mapOf("type" to "fetching")

    fun fetched(networks: List<String>): Map<String, Any?> =
        mapOf("type" to "fetched", "networks" to networks)

    fun trying(network: String, unitId: String, position: Int, total: Int): Map<String, Any?> =
        mapOf(
            "type" to "trying",
            "network" to network,
            "unitId" to unitId,
            "position" to position,
            "total" to total,
        )

    fun won(network: String, position: Int, latencyMs: Long): Map<String, Any?> =
        mapOf(
            "type" to "won",
            "network" to network,
            "position" to position,
            "latencyMs" to latencyMs,
        )

    fun lost(network: String, position: Int, reason: String): Map<String, Any?> =
        mapOf(
            "type" to "lost",
            "network" to network,
            "position" to position,
            "reason" to reason,
        )

    fun noFill(): Map<String, Any?> = mapOf("type" to "noFill")
}

File: ios/exelbid_plugin/Sources/exelbid_plugin/Mappers/WaterfallEventMapper.swift
import ExelBidSDK
import Foundation

/// Encodes an `ExelBidSDK.EBWaterfallEvent` into a Dart-friendly map sent under
/// the `"waterfall"` key of an `onWaterfall` event. The shape mirrors the
/// `EBWaterfallEvent` sealed class on the Dart side.
enum WaterfallEventMapper {
    static func encode(_ event: EBWaterfallEvent) -> [String: Any] {
        switch event {
        case .fetching:
            return ["type": "fetching"]
        case let .fetched(networks):
            return ["type": "fetched", "networks": networks]
        case let .trying(network, unitId, position, total):
            return [
                "type": "trying",
                "network": network,
                "unitId": unitId,
                "position": position,
                "total": total,
            ]
        case let .won(network, position, latency):
            return [
                "type": "won",
                "network": network,
                "position": position,
                "latencyMs": Int(latency * 1000),
            ]
        case let .lost(network, position, reason):
            return [
                "type": "lost",
                "network": network,
                "position": position,
                "reason": reasonString(reason),
            ]
        case .noFill:
            return ["type": "noFill"]
        }
    }

    private static func reasonString(_ reason: EBWaterfallFailReason) -> String {
        switch reason {
        case .adapterNotRegistered:
            return "adapterNotRegistered"
        case let .loadFailed(error):
            return "loadFailed: \(error.localizedDescription)"
        case let .timeout(seconds):
            return "timeout \(seconds)s"
        case .cancelled:
            return "cancelled"
        }
    }
}
